### ISSUE TITLE
Arch Linux: Atom build error due to dump-symbols.js

### ISSUE THREAD

**Issue by DAC324** [atom-community/atom#363]
Hello all,

trying to build Atom, I get
```
Transpiling Babel paths in /opt/.cache/yay/atom/src/atom-1.61.0-beta0/out/app
Transpiling CoffeeScript paths in /opt/.cache/yay/atom/src/atom-1.61.0-beta0/out/app
Transpiling CSON paths in /opt/.cache/yay/atom/src/atom-1.61.0-beta0/out/app
Transpiling PEG.js paths in /opt/.cache/yay/atom/src/atom-1.61.0-beta0/out/app
Generating module cache for /opt/.cache/yay/atom/src/atom-1.61.0-beta0/out/app
Generating pre-built less cache in /opt/.cache/yay/atom/src/atom-1.61.0-beta0/out/app/less-compile-cache
Generating metadata for /opt/.cache/yay/atom/src/atom-1.61.0-beta0/out/app/package.json
Generating API docs at /opt/.cache/yay/atom/src/atom-1.61.0-beta0/docs/output/atom-api.json
Dumping symbols in /opt/.cache/yay/atom/src/atom-1.61.0-beta0/out/symbols
/opt/.cache/yay/atom/src/atom-1.61.0-beta0/script/lib/dump-symbols.js:33
        throw new Error(error);
        ^

[Error: Error
  at /opt/.cache/yay/atom/src/atom-1.61.0-beta0/script/lib/dump-symbols.js:33:15
  at ChildProcess.<anonymous> (/opt/.cache/yay/atom/src/atom-1.61.0-beta0/script/node_modules/minidump/lib/minidump.js:27:7)
  at ChildProcess.emit (events.js:315:20)
  at maybeClose (internal/child_process.js:1021:16)
  at Socket.<anonymous> (internal/child_process.js:443:11)
  at Socket.emit (events.js:315:20)
  at Pipe.<anonymous> (net.js:674:12)
]
==> ERROR: A failure occurred in build().
    Aborting...
 -> error making: atom
```
This is due to the following code around line 33 of `src/atom-1.61.0-beta0/script/lib/dump-symbols.js`:
```
      if (error) {
        // fswin.node is only used on windows, ignore the error on other platforms
        if (process.platform !== 'win32' && binaryPath.match(/fswin.node/))
          return resolve();
       throw new Error(error);
```

**Comment by aminya** [comment]
It seems that you don't have the dependencies. See this for some of them:
https://github.com/atom-community/atom/blob/master/script/vsts/platforms/templates/preparation.yml

**Comment by DAC324** [comment]
That file lists the following dependencies:
```
build-essential (build-essentials)
ca-certificates 
xvfb (does not exist in Arch)
fakeroot 
git 
rpm (is this really needed on Arch Linux?)
libsecret-1-dev (libsecret)
libx11-dev (libx11)
libxkbfile-dev (libxkbfile)
xz-utils (xz)
xorriso (does not exist in Arch)
zsync (zsync in AUR)
libxss1 (libxss)
libgtk-3-0 (does not exist in Arch)
libasound2 (alsa)
libicu-dev (libicu50 in AUR)
software-properties-common (does not exist in Arch)
wget 
dpkg (is this really needed on Arch Linux?)
```
Strange, though that the previous `atom` version `1.58` was working without those. Which one has been introduced newly since then?

**Comment by DAC324** [comment]
Update:
Looks like this error comes from the way `atom` is built during creation of the AUR package. 
```
build() {
    cd "$_archive"

    pushd node_modules/git-utils
    env \
        npm_config_disturl=https://electronjs.org/headers \
        npm_config_runtime=$_electron \
        npm_config_target=$_electron_version \
        apm rebuild
    popd

    pushd script
    # Hack to avoid using Node > 12 (https://github.com/babel/babel/issues/11216)
    # Set ELECTRON_VERSION (see use-system-electron.patch)
    env \
        ELECTRON_RUN_AS_NODE=1 \
        ELECTRON_VERSION=$_electron_version \
        $_electron build # --no-bootstrap
}
```
If I only do 
```
git clone https://github.com/atom-community/atom.git
cd atom
script/bootstrap
```
as instructed at https://flight-manual.atom.io/hacking-atom/sections/hacking-on-atom-core/#platform-linux, the build succeeds.

**Comment by DAC324** [comment]
> It seems that you don't have the dependencies. See this for some of them: https://github.com/atom-community/atom/blob/master/script/vsts/platforms/templates/preparation.yml

Did some research, and I doubt that this error occurs due to missing dependencies. 
As per instructions, I have all of them in place:
```
sudo pacman -S --needed gconf base-devel git nodejs npm libsecret python libx11 libxkbfile
```
The problem seems to be an error in checking this condition:
```
       // fswin.node is only used on windows, ignore the error on other platforms
        if (process.platform !== 'win32' && binaryPath.match(/fswin.node/))
          return resolve();
```
Here, `resolve()` is only returned if `platform !== 'win32' and `binaryPath.match(/fswin.node/)` returns true (i.e., the path exists. As `fswin.node`, however, is used on Windows only, `binaryPath.match(/fswin.node/)` is supposed to return **false** on Linux. 

Hence, if I am not mistaken, the statement should read like this:
```
       // fswin.node is only used on windows, ignore the error on other platforms
        if (process.platform !== 'win32' && !binaryPath.match(/fswin.node/))
          return resolve();
```

**Comment by DAC324** [comment]
Update: 
For some weird reasons, 
```
        if (process.platform !== 'win32' && binaryPath.match(/fswin.node/))
```
never becomes true and consequently, an error is thrown. The culprit seems to be `binaryPath.match(/fswin.node/)` - no matter what it returns, the statement always remains false. 
For now, I therefore removed it
```
      if (error) {
        // fswin.node is only used on windows, ignore the error on other platforms
        if (process.platform !== 'win32')
          return resolve();
       throw new Error(error);
```
and with that change, `atom` can be built and is running.

**Comment by aminya** [comment]
You are building Atom on Arch, which is not supported. Please use something debian based. I suspect the issue is with some arch binaries. 

Regarding this code, this was added because `fswin.node` only works on Windows. 
https://github.com/atom-community/atom/blob/8625193dd21a6783550c2e491232f6b36b41afd8/script/lib/dump-symbols.js#L31-L32

**Comment by DAC324** [comment]
> Yes, you can change it to ignore dumping symbols on non-windows altogether, but you should see what the actual error is.

Of course, but unfortunately, `throw new Error(error);` does not seem to print out the contents of `error` as this variable seems to be a Boolean only. 
All you get as a result is
```
Dumping symbols in /opt/.cache/yay/atom/src/atom-1.63.0-dev/out/symbols
/opt/.cache/yay/atom/src/atom-1.63.0-dev/script/lib/dump-symbols.js:33
        throw new Error(error);
        ^

[Error: Error
  at /opt/.cache/yay/atom/src/atom-1.63.0-dev/script/lib/dump-symbols.js:33:15
  at ChildProcess.<anonymous> (/opt/.cache/yay/atom/src/atom-1.63.0-dev/script/node_modules/minidump/lib/minidump.js:27:7)
  at ChildProcess.emit (events.js:315:20)
  at maybeClose (internal/child_process.js:1021:16)
  at Socket.<anonymous> (internal/child_process.js:443:11)
  at Socket.emit (events.js:315:20)
  at Pipe.<anonymous> (net.js:674:12)
]
```
What I do not entirely understand is that it does not seem to matter what `binaryPath.match(/fswin.node/)` returns. 
If I am not mistaken, in this statement 
```
if (process.platform !== 'win32' && binaryPath.match(/fswin.node/))
          return resolve();
```
there should be a `return resolve()` if `process.platform !== 'win32' AND `binaryPath.match(/fswin.node/)` returns true (which seems to be in contrast to what this part of code is supposed to test). 
But anyway, the error is also thrown if I change the statement to
``` 
if (process.platform !== 'win32' && !binaryPath.match(/fswin.node/))
          return resolve();
```
Apparently, that `if` clause never becomes true as long as `binaryPath.match(/fswin.node/)` is included. I must admit that my current programming knowledge does not allow me to understand that.

**Comment by aminya** [comment]
Do you want to change the condition to only check for Windows?

**Comment by DAC324** [comment]
> Do you want to remove the condition? We can see what platforms exactly fail this in our CI

Indeed, according to 
```
 // fswin.node is only used on windows, ignore the error on other platforms
```
it appears sufficient if we check for `process.platform` **or** `binaryPath.match(/fswin.node/)`. 
Consequently, I [elsewhere read the suggestion to change the joining in that statement from `&&` to `||`](https://github.com/atom/atom/issues/25518#issuecomment-1154254508). I tested that already, and it seems to work as well. 
`   if (error) {
        // fswin.node is only used on windows, ignore the error on other platforms
        if (process.platform !== 'win32' || binaryPath.match(/fswin.node/))
          return resolve();
       throw new Error(error);`

**Comment by DeeDeeG** [comment]
Seems to be fixed by updating `minidump` to the latest version, in `script/package.json`. 0.22.0 --> 0.23.0

<details>

I tested this in a fresh install of the latest [Manjaro Linux](https://manjaro.org/), which is a type of Arch Linux. I got the same error situation as @DAC324 reported.

I tried debugging it. `dump_syms` had blank output on stdout and stderr, so that is why there was no detailed or helpful error message being printed. The binary being spawned really had no output. I think it's fixed by this upstream commit of breakpad? ["avoid dump_syms crashing if selected arch is not found"](https://chromium.googlesource.com/breakpad/breakpad/+/c85eb4a59b618f3beaad5445ceb1f865ffa8efdf)... Maybe.

Something about the way `dump_syms` exited made it hard to catch and handle properly, I guess. So it "leaked through" and halted the build script, even though we're supposed to be handling (ignoring) the error. 

In any case, it works for me now with updated `minidump`.

</details>
